src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/BasicSpringDataRepositoryDatabaseService.java

```diff
@@ -37,105 +42,189 @@
-                .collect(Collectors.toSet()));
+    public ModelWrapper<MODEL, String, ACCESSION> findAccessionVersion(ACCESSION accession, int version)
+            throws AccessionDoesNotExistException, AccessionDeprecatedException, AccessionMergedException {
+        ACCESSION_ENTITY result = doFindAccessionVersion(accession, version);
+        return toModelWrapper(result);
+    }
+
+    private ACCESSION_ENTITY doFindAccessionVersion(ACCESSION accession, int version) throws
+            AccessionDoesNotExistException, AccessionMergedException, AccessionDeprecatedException {
+        ACCESSION_ENTITY result = repository.findByAccessionAndVersion(accession, version);
+        if (result == null) {
+            assertAccessionMergedOrDeprecated(accession);
+            throw new AccessionDoesNotExistException(accession.toString(), version);
+        }
+        return result;
     }
 
     @Override
-    public AccessionWrapper<MODEL, String, ACCESSION> update(AccessionWrapper<MODEL, String, ACCESSION> accession)
-            throws AccessionDoesNotExistException, HashAlreadyExistsException {
-        Collection<ENTITY> accessionedElements = repository.findByAccession(accession.getAccession());
-        assertAccessionExists(accession, accessionedElements);
-        assertHashDoesNotExist(accession);
+    public void insert(List<ModelWrapper<MODEL, String, ACCESSION>> objects) {
+        Set<ACCESSION_ENTITY> entitySet = objects.stream().map(toEntityFunction).collect(Collectors.toSet());
+        repository.insert(entitySet);
+    }
 
-        int version = 1;
-        for (ENTITY accessionedElement : accessionedElements) {
-            if (version <= accessionedElement.getVersion()) {
-                version = accessionedElement.getVersion() + 1;
+    @Override
+    public AccessionWrapper<MODEL, String, ACCESSION> patch(ModelWrapper<MODEL, String, ACCESSION> accession)
+            throws AccessionDoesNotExistException, HashAlreadyExistsException, AccessionDeprecatedException,
+            AccessionMergedException {
+        List<ACCESSION_ENTITY> entities = getAccession(accession.getAccession());
+        int maxVersion = 1;
+        for (ACCESSION_ENTITY entity : entities) {
+            if (entity.getVersion() >= maxVersion) {
+                maxVersion = entity.getVersion();
             }
         }
+        accession.setVersion(maxVersion + 1);
+        repository.insert(Arrays.asList(toEntityFunction.apply(accession)));
+        return findAccession(accession.getAccession());
+    }
 
-        AccessionWrapper<MODEL, String, ACCESSION> newAccessionVersion =
-                new AccessionWrapper<>(accession.getAccession(), accession.getHash(), accession.getData(), version);
-        insert(Arrays.asList(newAccessionVersion));
-        return newAccessionVersion;
+    /**
+     * @param accessionId
+     * @return All entries of the last version of an accession. It is never empty
```

Comment:
I am a bit confused with this comment. According to the method implementation I think it is returning all versions of an accession, right?